### PR TITLE
Fix: Entities reducer refactor to enable update of redux-actions

### PR DESCRIPTION
**Pull request by Bouncey**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [ ] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX

#### Description

This PR supersedes #11754 and #11781 

`entities-reducer.js` required a refactor to accept new versions of `redux-actions`, see #11754 for more details.

I have two commits
- `entities-reducer` refactor
- `redux-actions` update

### CODE REVIEW

File: common/app/redux/entities-reducer.js [hunk 1]
```diff
@@ -1,103 +1,94 @@
...
+      ...action.meta.entities
     };
   }
   return state;
 }
+
+export default function entitiesReducer(state, action) {
+  const newState = metaReducer(state, action);
+  const user = userReducer(state && state.user || {}, action);
+  if (newState.user !== user && Object.keys(user).length > 0) {
```
Comment (BerkeleyTrue) [inline common/app/redux/entities-reducer.js]: Why are you doing the additional check for user keys? If the user ref values are not the same, that alone should signify that the user object has been updated and the overall state object should receive a  new ref value.

Comment (Bouncey) [inline common/app/redux/entities-reducer.js]: I was running into a problem `user` was just an empty Object. Without the `Object.keys` check, `state.user` was being totally overridden with an empty Object.

File: common/app/redux/actions.js [hunk 2]
```diff
@@ -92,7 +92,7 @@ export const updateUserEmail = createAction(
 // updateUserLang(username: String, lang: String) => Action
 export const updateUserLang = createAction(
   types.updateUserLang,
-  (username, lang) => ({ username, lang })
+  (username, lang) => ({ username, languageTag: lang })
```
Comment (BerkeleyTrue) [inline common/app/redux/actions.js]: Great catch!

File: common/app/redux/entities-reducer.js (hunk 1 above)
Comment (BerkeleyTrue) [inline common/app/redux/entities-reducer.js]: I see. After some debugging, I think I have found the reason.

Normally when combining reducers subsequent reducers should see the latest state, but if `metaReducer` makes a change to `state.user` when `userReducer` ends up seeing stale data. You should always feed `newState.user` to `userReducer`. Then you will not need the check for keys.

File: common/app/redux/entities-reducer.js [hunk 3]
```diff
@@ -1,103 +1,94 @@
...
+      ...state,
+      ...action.meta.entities
     };
   }
   return state;
 }
+
+export default function entitiesReducer(state, action) {
+  const newState = metaReducer(state, action);
+  const user = userReducer(state && state.user || {}, action);
```
Comment (BerkeleyTrue) [inline common/app/redux/entities-reducer.js]: initialState is already given to the `handleActions`. No need add the extra `|| {}` check.

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by BerkeleyTrue** [comment]
Looking good @Bouncey! Almost there

**Comment by Bouncey** [comment]
Passing `newState.user` helped greatly with the blank object. I also found a few typo and assignment bugs which caused a blank object to be returned along with the user object.

**Comment by Bouncey** [comment]
The build fails on the challenge solution for `Friendly Dates Ranges` because 2017. 

I will raise an issue to rectify

**Comment by Bouncey** [comment]
Please see #12305

**Comment by BerkeleyTrue** [comment]
Please rebase off of latest staging. That should fix the test issues

File: common/app/redux/entities-reducer.js [hunk 4]
```diff
@@ -1,103 +1,100 @@
...
         }
-      }
+    })
+  },
+  initialState.user
+);
+
+function metaReducer(state = initialState, action) {
+  if (action.meta && action.meta.entities) {
+    console.log('meta', action.meta.entities.user);
```
Comment (BerkeleyTrue) [inline common/app/redux/entities-reducer.js]: remove console.log

File: common/app/redux/entities-reducer.js [hunk 5]
```diff
@@ -1,103 +1,100 @@
...
+      ...state,
+      ...action.meta.entities
     };
   }
   return state;
 }
+
+export default function entitiesReducer(state, action) {
+  const newState = metaReducer(state, action);
+  console.log(newState.user);
```
Comment (BerkeleyTrue) [inline common/app/redux/entities-reducer.js]: remove console.log

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Review by BerkeleyTrue — APPROVED** [review]

**Pull request merged**